swarm/src/lib.rs

```diff
@@ -300,11 +301,24 @@ where
     pending_event: Option<(PeerId, PendingNotifyHandler, THandlerInEvent<TBehaviour>)>,
 }
 
-impl<TBehaviour> Unpin for Swarm<TBehaviour> where TBehaviour: NetworkBehaviour {}
+#[cfg(feature = "tokio")]
+pub type TokioSwarm<'a, TBehaviour> = Swarm<TBehaviour, libp2p_core::TokioExecutor<'a>>;
 
-impl<TBehaviour> Swarm<TBehaviour>
+#[cfg(feature = "async-std")]
+pub type AsyncStdSwarm<TBehaviour> = Swarm<TBehaviour, libp2p_core::AsyncStdExecutor>;
+
+impl<TBehaviour, TExecutor> Unpin for Swarm<TBehaviour, TExecutor>
 where
     TBehaviour: NetworkBehaviour,
+    TExecutor: Executor,
+{
+}
+
+impl<TBehaviour, TExecutor> Swarm<TBehaviour, TExecutor>
+where
+    TBehaviour: NetworkBehaviour,
+    TExecutor: Executor,
+    PoolConfig<TExecutor>: Default,
 {
     /// Builds a new `Swarm`.
     pub fn new(
```

Comment:
I am leaning towards deprecating this, defaulting to `Threadpool` in here (if construction fails, falling back to local spawns, same as now) and adding the following functions:

```rust
let swarm = Swarm::with_executor(transport, behaviour, local_peer_id, executor);
let swarm = Swarm::with_tokio_executor(transport, behaviour, local_peer_id);
let swarm = Swarm::with_async_std_executor(transport, behaviour, local_peer_id);
let swarm = Swarm::with_threadpool_executor(transport, behaviour, local_peer_id);
let swarm = Swarm::without_executor(transport, behaviour, local_peer_id);
```

Reply:
I would avoid the `without_executor`. It might be ambiguous.

Reply:
How do you create a `Swarm` then that runs its connections on the current task?

Reply:
Good question. Maybe with `Swarm::with_current_thread_executor`?

Reply:
Decided otherwise, I hope you like my solution.